### PR TITLE
MIGraphX support for 3-D and 1-D convolutions

### PR DESCRIPTION
**Pull request by pcf000**
Pull request #1505 generalised the convolution code to handle N dimensions (for the most part).  This pull request covers the conversion from migraphx through tosa to rock.  There are a couple bits that aren't completely general yet, but it does handle (with a patch to AMDMIGraphX) "make check" and the python_3dunet example.

### CODE REVIEW

**Comment by pcf000** [comment]
That "Changing 'rock.conv2d'" commit came from the rebase-via-merge process, and contains merge conflicts.  Confusing, though, since an earlier commit exists with the same summary but more extensive contents.

File: mlir/lib/CAPI/Dialect/MIGraphX.cpp
Comment (pcf000) [inline mlir/lib/CAPI/Dialect/MIGraphX.cpp]: MIGraphX likes environment variables to control things.  Also, there's no place to hang --mlir-ir-print-all in a python script.  I needed this to see what rocMLIR was doing underneath python_3dunet.  enableIRPrinting() requires disableMultithreading().

File: mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp
Comment (pcf000) [inline mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp]: migraphx.convolution becomes tosa.conv2d or tosa.conv3d.  There is no tosa.conv1d, so here we expand those into conv2d by adding an extra x1 dimension to the tensors and some dummy values to the attributes.  There are tosa.reshapes on the inputs and another after the result.

File: mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp
Comment (pcf000) [inline mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp:325]: Attributes for the dummy dimension.

File: mlir/lib/Conversion/TosaToRock/TosaToRock.cpp
Comment (pcf000) [inline mlir/lib/Conversion/TosaToRock/TosaToRock.cpp]: Bigger string for layouts with more dimensions.  I don't like this much, but I didn't find the correct ArrayRef or something else to generalise it.

File: mlir/lib/Conversion/TosaToRock/TosaToRock.cpp
Comment (pcf000) [inline mlir/lib/Conversion/TosaToRock/TosaToRock.cpp:155]: Really not "> 4" as much as "== 5";  ie, not general.  We also have mixed-mode layouts here, with x/y/h/w for 2-D and 0/1/2 for 3-D.  Some existing rocMLIR tests failed if I used "k01c" and I didn't want to spend the time to track it down.  Will need a cleanup ticket later.

File: mlir/lib/Conversion/TosaToRock/TosaToRock.cpp
Comment (pcf000) [inline mlir/lib/Conversion/TosaToRock/TosaToRock.cpp]: Begging for abstraction.  Conv3DConverter is nearly identical to ConvConverter, except for the op type and the sizes of biasShape and reassociations.

File: mlir/lib/Conversion/TosaToRock/TosaToRock.cpp
Comment (pcf000) [inline mlir/lib/Conversion/TosaToRock/TosaToRock.cpp]: While debugging, I wandered down the address-sanitiser path.  Didn't find my problem that way, but I did see that nodes erased here were sometimes used after erasure.  Collect them and then erase at the end.

File: mlir/lib/Dialect/Rock/Transforms/ConvToGemm.cpp
Comment (pcf000) [inline mlir/lib/Dialect/Rock/Transforms/ConvToGemm.cpp]: I had situations that wouldn't find the expected element in the toLayoutIdxs, causing the ->getSecond() to barf obscurely.

File: mlir/lib/Dialect/Rock/Tuning/RockTuningImpl.cpp
Comment (pcf000) [inline mlir/lib/Dialect/Rock/Tuning/RockTuningImpl.cpp]: No change here, just formatting that I'll fix.

File: mlir/lib/Dialect/Rock/Tuning/RockTuningImpl.cpp
Comment (pcf000) [inline mlir/lib/Dialect/Rock/Tuning/RockTuningImpl.cpp]: If we don't find a matching op, newSpace is deleted and then returned, causing additional obscure barfs.  A proposed improvement is to change the interface so this function can return nullptr and the caller can detect and handle the situation.

File: mlir/lib/Dialect/Rock/Tuning/RockTuningImpl.cpp
Comment (pcf000) [inline mlir/lib/Dialect/Rock/Tuning/RockTuningImpl.cpp]: Code below relies on the layout strings being the correct length so it can overwrite characters.  This code doesn't handle 1-D, which migraphx doesn't care about but other paths might.

File: external/llvm-project/mlir/include/mlir/Dialect/Tosa/IR/TosaOps.td
Comment (krzysz00) [inline external/llvm-project/mlir/include/mlir/Dialect/Tosa/IR/TosaOps.td]: We'll want to get this into a nice `[DO NOT SQUASH]` for for upstreaming

File: mlir/lib/CAPI/Dialect/MIGraphX.cpp
Comment (krzysz00) [reply mlir/lib/CAPI/Dialect/MIGraphX.cpp]: Ah. This is a slightly different bit of debug machinery though it is still meaningfully a subset of ticket 1227 (the debug flags one).

But also ... LLVM's got a utility for getenv, we might could use that here instead

File: mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp
Comment (krzysz00) [reply mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp]: Nit: no need to take `loc` by reference

File: mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp
Comment (krzysz00) [reply mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp:325]: Could you hang an attribute on the tosa.conv2d that's something like `rock.actually_1d`

File: mlir/lib/Conversion/TosaToRock/TosaToRock.cpp
Comment (krzysz00) [reply mlir/lib/Conversion/TosaToRock/TosaToRock.cpp]: Yeah, this should be `MutableArrayRef<char>`, probably

File: mlir/lib/Conversion/TosaToRock/TosaToRock.cpp
Comment (krzysz00) [inline mlir/lib/Conversion/TosaToRock/TosaToRock.cpp]: While we're here, none of these attributes need to be by const reference

File: mlir/lib/Conversion/TosaToRock/TosaToRock.cpp
Comment (krzysz00) [reply mlir/lib/Conversion/TosaToRock/TosaToRock.cpp]: Yeah, integer template, no?

Comment (krzysz00) [reply mlir/lib/Conversion/TosaToRock/TosaToRock.cpp]: Or constructor arguments

File: mlir/lib/Conversion/TosaToRock/TosaToRock.cpp
Comment (krzysz00) [inline mlir/lib/Conversion/TosaToRock/TosaToRock.cpp]: Nit: much less auto

File: mlir/lib/Conversion/TosaToRock/TosaToRock.cpp
Comment (krzysz00) [inline mlir/lib/Conversion/TosaToRock/TosaToRock.cpp]: Why?

File: mlir/lib/Dialect/Rock/Transforms/ConvToGemm.cpp
Comment (krzysz00) [reply mlir/lib/Dialect/Rock/Transforms/ConvToGemm.cpp]: Put a string on the assert?

File: mlir/lib/Dialect/Rock/Tuning/RockTuningImpl.cpp
Comment (krzysz00) [reply mlir/lib/Dialect/Rock/Tuning/RockTuningImpl.cpp]: This is `emitFatalError()` or such more than unreachable

File: mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp
Comment (krzysz00) [inline mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp]: `SmallVector`

File: mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp
Comment (krzysz00) [inline mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp]: `ArrayRef<int64_t> origShape`

File: mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp
Comment (krzysz00) [inline mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp]: Don't assert. `return emitError()` - this is a user error, not a failure of ours.

File: mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp
Comment (krzysz00) [inline mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp]: ... this is just `expShape.push_back(1);`

File: mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp
Comment (krzysz00) [inline mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp]: This is not `llvm_unreachable`, this is returning an error.

File: mlir/lib/Conversion/TosaToRock/TosaToRock.cpp
Comment (krzysz00) [reply mlir/lib/Conversion/TosaToRock/TosaToRock.cpp]: This is still `MutableArrayRef<char>`

File: mlir/lib/Conversion/TosaToRock/TosaToRock.cpp
Comment (krzysz00) [reply mlir/lib/Conversion/TosaToRock/TosaToRock.cpp]: Same not on not naking attributes by const reference

File: mlir/lib/Conversion/TosaToRock/TosaToRock.cpp
Comment (krzysz00) [inline mlir/lib/Conversion/TosaToRock/TosaToRock.cpp]: Let's go and make this 8 for good stack alignment

File: mlir/lib/Conversion/TosaToRock/TosaToRock.cpp
Comment (krzysz00) [reply mlir/lib/Conversion/TosaToRock/TosaToRock.cpp:155]: What tests, and failed how?

Comment (krzysz00) [inline mlir/lib/Conversion/TosaToRock/TosaToRock.cpp:155]: Could we use that `rock.actually_conv_1d` attribute here to drop those dummy dimensions?

File: mlir/lib/Conversion/TosaToRock/TosaToRock.cpp
Comment (krzysz00) [reply mlir/lib/Conversion/TosaToRock/TosaToRock.cpp]: Yeah no, let's actually debug that - but maybe in a future ticket

File: mlir/lib/Dialect/Rock/Transforms/ConvToGemm.cpp
Comment (krzysz00) [inline mlir/lib/Dialect/Rock/Transforms/ConvToGemm.cpp]: Commented-out code?

File: mlir/lib/Dialect/Rock/Transforms/ConvToGemm.cpp
Comment (krzysz00) [inline mlir/lib/Dialect/Rock/Transforms/ConvToGemm.cpp:459]: Nice catch!

File: mlir/lib/Dialect/Rock/Transforms/ConvToGemm.cpp
Comment (krzysz00) [reply mlir/lib/Dialect/Rock/Transforms/ConvToGemm.cpp]: ... Heck, this can be triggered by user input, it should be a `return failure()` instead

File: mlir/lib/Dialect/Rock/Tuning/RockTuningImpl.cpp
Comment (krzysz00) [reply mlir/lib/Dialect/Rock/Tuning/RockTuningImpl.cpp]: ... you could instead `SmallString<8> oLayout(size, '#');` above - I might have the constructor backwards

File: mlir/lib/Dialect/Rock/Tuning/RockTuningImpl.cpp
Comment (krzysz00) [inline mlir/lib/Dialect/Rock/Tuning/RockTuningImpl.cpp]: This creates way too many `#`s also

File: mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp
Comment (pcf000) [reply mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp]: Actually, no.  It inserts the 1 as the next-to-last element.  I didn't want to mess with channels, just add a fake "w".

Comment (krzysz00) [reply mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp]: Ah, right, Tosa's NHWC

Then let's construct the shape by parts - a small-vector of everything but channels (via `drop_back()`), add a `1`, and then push the channels on

File: mlir/lib/Conversion/TosaToRock/TosaToRock.cpp
Comment (pcf000) [reply mlir/lib/Conversion/TosaToRock/TosaToRock.cpp:149]: On further study, I can't make these SmallStrings into MutableArrayRef in expandTensor(), because part of the expansion is to insert a "g" into the layout string.  As a MutableArrayRef, I can modify the existing elements, but since the buffer is owned by somebody else, I can't expand it.  Therefore, SmallString<8> all around.

File: mlir/lib/Conversion/TosaToRock/TosaToRock.cpp
Comment (pcf000) [reply mlir/lib/Conversion/TosaToRock/TosaToRock.cpp:155]: Yes, but I'd prefer to leave that to the hypothetical cleanup pull-request, in the interests of not breaking anything during the review close to the deadline.

File: mlir/lib/Conversion/TosaToRock/TosaToRock.cpp
Comment (pcf000) [reply mlir/lib/Conversion/TosaToRock/TosaToRock.cpp]: On further study, I can't make these SmallStrings into MutableArrayRef in expandTensor(), because part of the expansion is to insert a "g" into the layout string. As a MutableArrayRef, I can modify the existing elements, but since the buffer is owned by somebody else, I can't expand it. Therefore, SmallString<8> all around.

File: external/llvm-project/mlir/include/mlir/Dialect/Tosa/IR/TosaOps.td
Comment (manupak) [reply external/llvm-project/mlir/include/mlir/Dialect/Tosa/IR/TosaOps.td]: We have lots of `if (group > 1 && isConv2DOp` added to external in TosaToLinalgNamed lowering that has not been upstreamed.
Therefore, I d like to request a MIGraphX e2e test with `--clone-harness` make sure this change is properly tested -- otherwise we ll have a hard time verifying fused modules from MIGraphX.

File: external/llvm-project/mlir/include/mlir/Dialect/Tosa/IR/TosaOps.td
Comment (manupak) [inline external/llvm-project/mlir/include/mlir/Dialect/Tosa/IR/TosaOps.td]: please remove spurious unncessary changes to external/ which will affect upstream merges.

File: external/llvm-project/mlir/include/mlir/Dialect/Tosa/IR/TosaOps.td
Comment (manupak) [inline external/llvm-project/mlir/include/mlir/Dialect/Tosa/IR/TosaOps.td]: same here

File: external/llvm-project/mlir/lib/Conversion/GPUToROCDL/LowerGpuOpsToROCDLOps.cpp
Comment (manupak) [inline external/llvm-project/mlir/lib/Conversion/GPUToROCDL/LowerGpuOpsToROCDLOps.cpp]: Why this change here?

File: mlir/lib/CAPI/Dialect/MIGraphX.cpp
Comment (manupak) [inline mlir/lib/CAPI/Dialect/MIGraphX.cpp]: How is this different from using MIGRAPHX_MLIR_TRACE_MLIR?
If we are doing this change, I d like that documented here.

File: mlir/lib/CAPI/Dialect/MIGraphX.cpp
Comment (manupak) [inline mlir/lib/CAPI/Dialect/MIGraphX.cpp]: Please refactor this to a single function to reduce error surface.

File: mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp
Comment (manupak) [inline mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp:243]: [stylistic] nit : I generally would use a lambda if it interacts heavily with variables in the scope where as in this one only uses rewriter and loc. I d consider defining a function with proper return types and some docs.

File: mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp
Comment (manupak) [inline mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp]: please add this to RockAttrDefs and use here.

File: mlir/lib/Dialect/Rock/Tuning/RockTuningImpl.cpp
Comment (manupak) [inline mlir/lib/Dialect/Rock/Tuning/RockTuningImpl.cpp]: nit: auto usage is only encouraged if its obvious as in the previous line in llvm/mlir project.
please use the type and expressive name here.

**Review by manupak — DISMISSED** [review]
I think the tosa -> linalg plumbing is not done here and it is required if you are changing tosa. I d like that verified as a clone test.

[non blocking on this PR because we have already added 'group' but have not made an attempt of upstreaming]  As a secondary point, I d like at least raising the point of adding group dimension in the TOSA spec discourse that is a pre-requisite for upstreaming here : https://discuss.mlplatform.org/c/discussion-of-the-tosa-specification/5

File: external/llvm-project/mlir/include/mlir/Dialect/Tosa/IR/TosaOps.td
Comment (pcf000) [reply external/llvm-project/mlir/include/mlir/Dialect/Tosa/IR/TosaOps.td]: What is involved in making such a test?  (I've never committed anything to MIGraphX.)  Would it be better to similarly test for "group > 1" in the migraphx conversion and leave the grouped ones out of the rock.conv path?

File: external/llvm-project/mlir/include/mlir/Dialect/Tosa/IR/TosaOps.td
Comment (pcf000) [reply external/llvm-project/mlir/include/mlir/Dialect/Tosa/IR/TosaOps.td]: Trailing whitespace, the bane of my existence.

File: external/llvm-project/mlir/lib/Conversion/GPUToROCDL/LowerGpuOpsToROCDLOps.cpp
Comment (pcf000) [reply external/llvm-project/mlir/lib/Conversion/GPUToROCDL/LowerGpuOpsToROCDLOps.cpp]: I did not change this file initially.  I had a clang-format failure because of a whitespace change.  When I synced that, I had a test failure because this layout variable changed on the develop branch while I still had the old version.

File: mlir/lib/CAPI/Dialect/MIGraphX.cpp
Comment (pcf000) [reply mlir/lib/CAPI/Dialect/MIGraphX.cpp]: MIGRAPHX_TRACE_MLIR dumps the IR at a couple of selected spots.  This variable enables the mechanism behind --print-ir-after-all, which dumps the IR after each pass.  For me, it was the difference between seeing migraphx.convolution and seeing where tosa.conv3d became rock.conv.

Krzysztof's issue #1227 will make this variable unnecessary by enabling a mechanism to take command-line options through an environment variable.  I'll go ahead and remove this in anticipation.

File: mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp
Comment (pcf000) [reply mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp]: Thanks, I didn't know about that.

File: external/llvm-project/mlir/lib/Conversion/GPUToROCDL/LowerGpuOpsToROCDLOps.cpp
Comment (manupak) [reply external/llvm-project/mlir/lib/Conversion/GPUToROCDL/LowerGpuOpsToROCDLOps.cpp]: Yes I think if you merge the develop into your branch -- even when your branch dont have conflicts with develop -- this wont be needed because clang-format checks against origin/develop and if your branch is upto date with origin/develop clang-format wont even touch this file.

**Review by manupak — COMMENTED** [review]
Only few things to go.

## Non-test code :

1) Please merge the latest develop into this branch --> that would make the clang-format changes go away. Then this PR would not need external/ changes and [DO NOT SQUASH]

## Test-code :

1) I d like at least a single e2e test. It should be trivial to add one. here is an example : https://github.com/ROCm/rocMLIR/blob/develop/mlir/test/fusion/pr-e2e/mixr-slice-convolution-add-small.mlir (note --clone-harness)
2) There are changes in TosaToRock -- there should be at least a single test in TosaToRock conversion.
3) There are changes in MIGraphXToTosa -  there should be at least a single test in MIGraphXToTosa conversion.

File: mlir/lib/CAPI/Dialect/MIGraphX.cpp
Comment (manupak) [reply mlir/lib/CAPI/Dialect/MIGraphX.cpp]: Cool thanks! I was going to suggest this passManager option setting should be done by MIGraphX who owns it.

**Comment by pcf000** [comment]
> 1. Please merge the latest develop into this branch --> that would make the clang-format changes go away. 

Pretty to think so, but we'll see.

> Then this PR would not need external/ changes and [DO NOT SQUASH]

Also because the tosa.conv3d group attribute is deferred, the original reason for it.

> 1. I d like at least a single e2e test. It should be trivial to add one. here is an example : 

I had been leaning on the migraphx "make check" tests, but this is fair.

> (note --clone-harness)

Ah.  I'd thought that was an migraphx thing.

**Review by manupak — APPROVED** [review]
Thanks! I do not have major concerns here.

**Pull request merged**
